html2docx.py

```diff
@@ -5,149 +5,132 @@
 
 from docx import Document
 from docx.text.paragraph import Paragraph
+from docx.text.run import Run
 
-WHITESPACE_RE = re.compile(r"\s+")
+PT_WHITESPACE = re.compile(r"\s+")
 
 
 class HTML2Docx(HTMLParser):
-    header_re = re.compile(r"^h[1-6]$")
-
     def __init__(self, title: str):
         super().__init__()
         self.doc = Document()
         self.doc.core_properties.title = title
 
-        self.p: Paragraph = None
-        self.runs: List[Run] = []
-        self.list_style: List[str] = []
-        self.bold = 0
-        self.italic = 0
-        self.underline = 0
-        self.extra_data = ""
-
-    def feed(self, data: str) -> None:
-        super().feed(data)
-        self.finish()
+        # Doc class members
+        self.p: Optional[Paragraph] = None
+        self.r: Optional[Run] = None
+
+        # Formatting options
+        self.style: Optional[str] = None
+        self.pre = False
+        self.list_styles: List[str] = []
+        self.attrs: List[List[str]] = []
+        self.collapse_space = True
+
+        self.href = ""
+
+    def _add_p(self, attrs: List[Tuple[str, Optional[str]]] = []) -> None:
+        style = self.list_styles and self.list_styles[-1] or self.style
+        self.p = self.doc.add_paragraph(style=style)
+        self.r = None
+        self.collapse_space = True
+        self.attrs = []
+
+    def _finish_p(self) -> None:
+        if self.r:
+            self.r.text = self.r.text.rstrip()
+        self.p = None
+        self.r = None
+
+    def _add_r(self, attrs: Optional[List[str]] = None) -> None:
+        if self.p is None:
+            self._add_p()
+        if attrs is not None:
+            self.attrs.append(attrs)
+        assert self.p is not None
+        self.r = self.p.add_run()
+        for run_style in (a for s in self.attrs for a in s):
```

Comment:
I think this would read better as nested for loops rather than generators inside loop statements.